### PR TITLE
Vectorized kll updates

### PR DESCRIPTION
**Pull request by leerho**
The primary focus of this PR is to add vectorized updates to the KLL Doubles Sketches.  
The changes and rationale by class is as follows:

KllDirectDoublesSketch
  - Grouped all the Min/Max methods together to make them easier to understand as they are closely related.
  - Added get Max/Min Internal methods. This is rather subtle. The public getMinItem() and getMaxItem() throw an exception if the sketch is empty. IsEmpty() is defined as getN() == 0 and when the sketch is empty during normal operation the max/min values are NaN. Which means that the state of the value of min/max must be in sync with the value of N. However, during an internal compaction or merge operation, these two values can get out-of-sync, making coding and debugging very difficult due to very sensitive sequencing of internal operations.  This problem was discovered during the implementation of the vector update methods, where these two values can very easily get out of sync. Adding these internal getters of Min/Max aleviates this problem.
  - The min/max methods themselves needed to be simplified.
  - The incN() is now incN(int).  required for vector updates.
  - new setDoubleItemsArrayAt(index, items[], srcOffset, length) method) for Vector updates.

KllDoublesSketch
  - Moved the Min/Max methods together
  - Moved the SortedView methods together
  - Moved all the update update methods together and grouped by regular, weighted and vector.
  - Moved the updateDouble(...) methods from KllDoublesHelper into this class. This allows for some collapsing of methods and simpler coding especially now that we have normal update, weighted update and vector update methods.
  
KllHeapDoublesSketch
  - Grouped all the Min/Max methods together as above
  - Added get Max/Min Internal methods as above.
  - The incN() is now incN(int) as above
  - new setDoubleItemsArrayAt(index, items[], srcOffset, length) method) as above 

KllDoublesHelper
  - Moved the updateDoubles(...) methods to KllDoublesSketch
  - Some modifications due to change in location of the updateDoubles(...) methods
  - These internal methods now refer to the new getMin/MaxItemInternal methods.

KllDirectCompactItemssketch
  - The incN() is now incN(int) as above
  
KllDirectFloatsSketch
  - The incN() is now incN(int) as above

KllFloatsHelper
  - The incN() is now incN(int) as above
  
KllHeapFloatsSketch
  - The incN() is now incN(int) as above
  
KllHeapItemsSketch
 - The incN() is now incN(int) as above
 
KllItemsHelper
 - The incN() is now incN(int) as above
 
KllSketch
 - The incN() is now incN(int) as above
 
TESTS

KllDoublesSketchTest
  - Added 3 extensive vector tests
  
KllDoublesSketchSerDeTest
  - Added some blank lines for better readability
  
KllDirectDoublesSketchTest
  - Added vector and weighted tests

### CODE REVIEW

File: src/main/java/org/apache/datasketches/kll/KllDoublesSketch.java [hunk 1]
```diff
@@ -336,12 +363,98 @@ public void update(final double item, final long weight) {
...
+   * @param items the vector of items
+   * @param offset the starting index of the items[] array
+   * @param length the number of items
+   */
+  public void update(final double[] items, final int offset, final int length) {
+    if (readOnly) { throw new SketchesArgumentException(TGT_IS_READ_ONLY_MSG); }
+    if (length == 0) { return; }
+    if (!hasNaN(items, offset, length)) {
+      updateDouble(items, offset, length); //fast path
+      return;
```
Comment (AlexanderSaydakov) [inline src/main/java/org/apache/datasketches/kll/KllDoublesSketch.java:392]: should this path invalidate the sorted view?

File: src/main/java/org/apache/datasketches/kll/KllDoublesSketch.java [hunk 2]
```diff
@@ -336,12 +363,98 @@ public void update(final double item, final long weight) {
...
+      setMinItem(min(getMinItemInternal(), srcItems[i]));
+      setMaxItem(max(getMaxItemInternal(), srcItems[i]));
+    }
+  }
+
+  // this returns on the first detected NaN.
+  private static boolean hasNaN(final double[] items, final int offset, final int length) {
+    final int end = offset + length;
+    for (int i = offset; i < end; i++) {
+      if (Double.isNaN(items[i])) { return true; } else { continue; }
```
Comment (AlexanderSaydakov) [inline src/main/java/org/apache/datasketches/kll/KllDoublesSketch.java:452]: unnecessary else

File: src/test/java/org/apache/datasketches/kll/KllDirectDoublesSketchTest.java [hunk 3]
```diff
@@ -644,6 +645,25 @@ public void checkMergeExceptionsWrongType() {
     try { sk2.merge(sk1); fail(); } catch (ClassCastException e) { }
   }
 
+  @Test
+  public void checkVectorUpdate() {
+    WritableMemory dstMem = WritableMemory.allocate(6000);
```
Comment (AlexanderSaydakov) [inline src/test/java/org/apache/datasketches/kll/KllDirectDoublesSketchTest.java:650]: do we care about "final" in tests?

File: src/main/java/org/apache/datasketches/kll/KllDoublesSketch.java (hunk 1 above)
Comment (leerho) [reply src/main/java/org/apache/datasketches/kll/KllDoublesSketch.java:392]: Ooh! Nice catch!  will fix.

File: src/main/java/org/apache/datasketches/kll/KllDoublesSketch.java (hunk 2 above)
Comment (leerho) [reply src/main/java/org/apache/datasketches/kll/KllDoublesSketch.java:452]: Yep! will fix.

File: src/main/java/org/apache/datasketches/kll/KllDoublesSketch.java [hunk 4]
```diff
@@ -336,12 +363,98 @@ public void update(final double item, final long weight) {
...
+      return;
+    }
+    //has at least one NaN
+    boolean oneUpdate = false;
+    final int end = offset + length;
+    for (int i = offset; i < end; i++) {
+      final double v = items[i];
+      if (!Double.isNaN(v)) {
+        updateDouble(this, v); //normal path
+        oneUpdate = true;
```
[the remark this answers is not in the document]
Comment (leerho) [reply src/main/java/org/apache/datasketches/kll/KllDoublesSketch.java:401]: Agreed. Will fix.

File: src/test/java/org/apache/datasketches/kll/KllDirectDoublesSketchTest.java (hunk 3 above)
Comment (leerho) [reply src/test/java/org/apache/datasketches/kll/KllDirectDoublesSketchTest.java:650]: I will fix for this file, but removing these in all the tests will be a cleanup task for later.

**Review by AlexanderSaydakov — APPROVED** [review]

**Pull request merged**